pkg/timeutil/periodic.go

```diff
@@ -0,0 +1,240 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package timeutil
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+type periodicStart struct {
+	dayOfWeek      int
+	hourOfDay      int
+	minuteOfHour   int
+	secondOfMinute int
+}
+
+type Periodic struct {
+	start    periodicStart
+	duration time.Duration
+}
+
+type Period struct {
+	Start time.Time
+	End   time.Time
+}
+
+// ParsePeriodic returns a Periodic specified as a start and duration.
+func ParsePeriodic(start, duration string) (pc Periodic, err error) {
+	if pc.start, err = parseStart(start); err != nil {
+		return pc, fmt.Errorf("unable to parse start: %v", err)
+	}
+	if pc.duration, err = time.ParseDuration(duration); err != nil {
+		return pc, fmt.Errorf("unable to parse duration: %v", err)
+	}
+	if pc.duration < time.Duration(0) {
+		return pc, fmt.Errorf("duration cannot be negative")
+	}
+	return
+}
+
+var weekdays = map[string]int{
+	"sun": int(time.Sunday),
+	"mon": int(time.Monday),
+	"tue": int(time.Tuesday),
+	"wed": int(time.Wednesday),
+	"thu": int(time.Thursday),
+	"fri": int(time.Friday),
+	"sat": int(time.Saturday),
+}
+
+// parseStart parses a string into a periodicStart.
+func parseStart(start string) (ps periodicStart, err error) {
+	ps.dayOfWeek = -1
+	f := strings.Fields(start)
+	nf := len(f)
+
+	if nf != 1 && nf != 2 {
+		return ps, fmt.Errorf("wrong number of fields")
+	}
+
+	if nf == 2 {
+		if dow, ok := weekdays[strings.ToLower(f[0])]; ok {
+			ps.dayOfWeek = dow
+		} else {
+			return ps, fmt.Errorf("invalid day of week: %q", f[0])
+		}
+		start = f[1]
+	}
+
+	n := 0
+	n, err = fmt.Sscanf(start, "%d:%d", &ps.hourOfDay, &ps.minuteOfHour)
+
+	// check Sscanf failure
+	if n != 2 || err != nil {
+		return ps, fmt.Errorf("invalid time of day: %q", start)
+	}
+
+	// check hour range
+	if ps.hourOfDay < 0 || ps.hourOfDay > 23 {
+		return ps, fmt.Errorf("invalid time of day: %q", start)
+	}
+
+	// check minute range
+	if ps.minuteOfHour < 0 || ps.minuteOfHour > 59 {
+		return ps, fmt.Errorf("invalid time of day: %q", start)
+	}
+
+	return
+}
+
+// DurationToStart returns the duration between the supplied time and the start
+// of Periodic's relevant period.
+// If we're in a period, a value <= 0 is returned, indicating how
+// deep into period we are.
+// If we're outside a period, a value > 0 is returned, indicating how long
+// before the next period starts.
+func (pc Periodic) DurationToStart(ref time.Time) time.Duration {
+	prev := pc.Previous(ref)
+	if prev.End.After(ref) || prev.End.Equal(ref) {
+		return prev.Start.Sub(ref)
+	}
+	return pc.Next(ref).Start.Sub(ref)
+}
+
+// Previous returns Periodic's previous Period occurrence relative to ref.
+func (pc Periodic) Previous(ref time.Time) (p Period) {
+	if pc.start.dayOfWeek != -1 { // Weekly
+		if pc.cmpDayOfWeek(ref) >= 0 {
+			// this week
+			p.Start = time.Date(ref.Year(),
```

Comment:
Is the only diff between all these is the Day field? it's crying out for a helper function, this function's difficult to parse